da/lazyledger/lazyledger.go

```diff
@@ -0,0 +1,168 @@
+package lazyledger
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"github.com/pelletier/go-toml"
+	"google.golang.org/grpc"
+
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/tx"
+	appclient "github.com/lazyledger/lazyledger-app/x/lazyledgerapp/client"
+	apptypes "github.com/lazyledger/lazyledger-app/x/lazyledgerapp/types"
+
+	"github.com/lazyledger/optimint/da"
+	"github.com/lazyledger/optimint/log"
+	"github.com/lazyledger/optimint/types"
+)
+
+type Config struct {
+	// PayForMessage related params
+	NamespaceID []byte
+	PubKey      []byte
+	BaseRateMax uint64 // currently not used
+	TipRateMax  uint64 // currently not used
+
+	// temporary fee fields
+	GasLimit  uint64
+	FeeAmount uint64
+
+	// RPC related params
+	RPCAddress string
+	ChainID    string
+	Timeout    time.Duration
+
+	// keyring related params
+
+	// KeyringAccName is the name of the account registered in the keyring
+	// for the `From` address field
+	KeyringAccName string
+	// Backend is the backend of keyring that contains the KeyringAccName
+	Backend string
+	RootDir string
+}
+
+type LazyLedger struct {
+	config Config
+	logger log.Logger
+
+	keyring keyring.Keyring
+
+	rpcClient *grpc.ClientConn
+}
+
+var _ da.DataAvailabilityLayerClient = &LazyLedger{}
+
+// Init is called once to allow DA client to read configuration and initialize resources.
+func (ll *LazyLedger) Init(config []byte, logger log.Logger) error {
+	ll.logger = logger
+	err := toml.Unmarshal(config, &ll.config)
+	if err != nil {
+		return err
+	}
+
+	// TODO(tzdybal): this means interactive reading from stdin - shouldn't we replace this somehow?
+	userInput := os.Stdin
+	ll.keyring, err = keyring.New(ll.config.KeyringAccName, ll.config.Backend, ll.config.RootDir, userInput)
+	return err
+}
```

Comment:
this is a super minor completely optional nit: perhaps this could be `NewLazyLedgerClient(keyring keyring.Keyring, config Config, logger log.Logger) *LazyLedger` or something similar?

Reply:
It's an implementation of [extremely generic interface](https://github.com/celestiaorg/optimint/blob/main/da/da.go#L30). This way it's possible to handle entire life-cycle of DA client uniformly in app.  

The entire point of this interface is to give maximum flexibility for DA client implementations (so even config is just a `[]byte` and client can interpret it as needed).